### PR TITLE
Tracking: Better support for ndarrays and tensors

### PR DESCRIPTION
**Pull request by ntakouris**
## What changes are proposed in this pull request?

See #5739 

Many times, you log metrics such as loss per step (tensors with gradients) or other numpy arrays (ndarrays). It would be very useful if there was built-in support to:

**For params**
- Extract single (`.item()`) or multiple values from ndarrays ([tolist](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tolist.html) ?)

**For metrics**
- [Detach torch tensors from the gradient graph](https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html) - pytorch lightning does this too
- Move tensors from gpu to cpu, convert to numpy if needed (ex. to convert to list)
- Extract values from single-element tensors or ndarrays, raise if resulting value is not supported (not float?)

Adding those few simple features, can make the usage much more intuitive and simpler, for deep learning engineers. Here's an example:

```py
anchors = np.array([[16], [32], [64], [128], [256]], dtype=np.int32)
x = np.array([1.0])

# in your train on batch func
loss = ...
```

- Before
```py
mlflow.log_params({'anchors': anchors.tolist(), 'x': x.item(), 'loss': loss.detach().item()/.tolist()})
````

- After
```py
mlflow.log_params({'anchors': anchors, 'x': x, 'loss': loss})
````
## How is this patch tested?

- WIP: Will add tests once some initial code review is performed
- 
## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

WIP

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/entities/conversion_utils.py
Comment (WeichenXu123) [inline mlflow/entities/conversion_utils.py:22]: Can we avoid importing pyspark/torch/tensorflow when we importing this `conversion_utils` module ? We can move these lines into function.

File: mlflow/entities/param.py
Comment (WeichenXu123) [inline mlflow/entities/param.py:13]: I would like to keep `if isinstance(key, pyspark.ml.param.Param):` and `key = key.name` logic inside `Param.__init__`, because convert_to_str_if_possible is a generic function but the `Param.name` logic only apply to "Param key" case.

Comment (WeichenXu123) [inline mlflow/entities/param.py:13]: I would like to keep `if isinstance(key, pyspark.ml.param.Param):` and `key = key.name` logic inside `Param.__init__`, because convert_to_str_if_possible is a generic function but the `Param.name` logic only apply to "Param key" case.

**Review by WeichenXu123 — COMMENTED** [review]
We need to document log_metric / log_param related APIs, to say we now support logging these types values.

File: mlflow/entities/param.py
Comment (ntakouris) [reply mlflow/entities/param.py:13]: The refactor which was performed leads to equivalent behaviour. Is there a specific reason that you want this branching inside the constructor even if it leads to the exact same result?

I have not checked if tests cover that because I am waiting for the code review to end.

File: mlflow/entities/conversion_utils.py
Comment (ntakouris) [reply mlflow/entities/conversion_utils.py:39]: +1, will do.

I think we need to also introduce a `MetricKey` and `MetricValue` type, so that:

```py
MetricKey = Union[str, pyspark.ml.param.Param]
MetricValue = Union[float, np.ndarray, tf.Tensor, torch.Tensor]
```

... but I am not sure how to do that without eagerly importing all these modules (which I guess we want to avoid)

Would just adding a doc, naming and tests should be sufficient?

File: mlflow/entities/conversion_utils.py
Comment (WeichenXu123) [reply mlflow/entities/conversion_utils.py:22]: You don't need to use flag like `__np_module_imported`

if we execute `import xxx` command twice, the second execution do nothing, because when we import one module, it is cached in `sys.modules`.

File: mlflow/entities/param.py
Comment (WeichenXu123) [reply mlflow/entities/param.py:13]: ok it's fine.

File: mlflow/entities/conversion_utils.py
Comment (WeichenXu123) [reply mlflow/entities/conversion_utils.py:39]: Don't need to add type hint for now. Avoid importing unnecessary modules is more important. The type hint requires importing modules in top level it will cause the issue.

**Comment by WeichenXu123** [comment]
@ntakouris Coud you rerun failed CI and see whether they passed ?

**Comment by ntakouris** [comment]
@WeichenXu123 can give this a final review? tests pass locally.

File: examples/h2o/random_forest.ipynb
Comment (dbczumar) [inline examples/h2o/random_forest.ipynb:94]: @ntakouris Do we need to make these formatting changes to the example notebooks? If not, can we revert them? Let me know if I'm missing something, but these don't appear to be related to the PR.

File: mlflow/entities/metric.py
Comment (dbczumar) [inline mlflow/entities/metric.py:15]: @ntakouris Apologies if I'm asking a question that's already been answered, but can you provide some information about the use case for having numpy arrays / tensors as metric names? This seem useful as metric values, though MLflow only supports scalar metrics; however, I don't see a clear use case for these as metric names (`key` == `name`). Also, as far as I'm aware, MLflow's Python APIs will cast metric names to strings automatically during logging, so I don't think this is strictly necessary.

File: mlflow/entities/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/entities/metic_value_conversion_utils.py:88]: This will throw if the array has multiple elements. What does the user-facing exception look like in that case? Should we catch the exception and throw something more readable?

File: mlflow/entities/metric.py
Comment (dbczumar) [inline mlflow/entities/metric.py:13]: @ntakouris Rather than making changes to the low-level metric entity, I think we should update the fluent `mlflow.log_metric` and `mlflow.log_metrics()` APIs to support these additional types (https://github.com/mlflow/mlflow/blob/37c678320fbf82b93d2614dd14c0bee784d95175/mlflow/tracking/fluent.py#L559-L587). The fundamental definition of a metric should have a string key and a scalar numeric value, but upstream user-facing APIs can definitely accept multiple types and coerce them to this format. Please let me know if you have questions here.

**Review by dbczumar — CHANGES_REQUESTED** [review]
@ntakouris Thank you for raising this PR! Supporting scalar numpy arrays / tensors for metric logging seems very convenient for users. I left some feedback about the level of abstraction where we're introducing this capability as well as some questions about the use case for arrays as keys, which seems a bit less clear. Please let me know if you have questions! Looking forward to getting this merged!

cc @WeichenXu123

File: mlflow/entities/metic_value_conversion_utils.py
Comment (ntakouris) [reply mlflow/entities/metic_value_conversion_utils.py:88]: This will just raise a `ValueError`, because it expects this to be a singular value. I can maybe catch that value error and re-raise a `ValueError` with a more helpful message?

File: examples/h2o/random_forest.ipynb
Comment (ntakouris) [reply examples/h2o/random_forest.ipynb:94]: My apologies, I thought I reverted that, but it certainly did not take effect. It's was a sideffect of using `black . `

File: mlflow/entities/metric.py
Comment (ntakouris) [reply mlflow/entities/metric.py:13]: Agreed. Will migrate changes to the high level api.

File: mlflow/entities/metric.py
Comment (ntakouris) [reply mlflow/entities/metric.py:15]: I think we shouldn't do that for metric names. For param values though, we can convert ndarrays/tensors to str though.

File: mlflow/entities/metic_value_conversion_utils.py
Comment (WeichenXu123) [reply mlflow/entities/metic_value_conversion_utils.py:88]: Shall we catch the original exception and wrap it to be a `MlflowException` ?

**Comment by ftsamis** [comment]
LGTM

**Comment by ntakouris** [comment]
@dbczumar most changes are done, do you want to include them into `mlflow/tracking/_tracking_service/client.py` , `mlflow/tracking/fluent.py` or `mlflow/tracking/client.py` ?

File: examples/h2o/random_forest.ipynb
Comment (dbczumar) [reply examples/h2o/random_forest.ipynb:94]: Got it! Can we revert these by checking out the notebooks from `master`?

**Comment by dbczumar** [comment]
> @dbczumar most changes are done, do you want to include them into `mlflow/tracking/_tracking_service/client.py` , `mlflow/tracking/fluent.py` or `mlflow/tracking/client.py` ?

@ntakouris Apologies for the delay. Can we add the changes to `mlflow/tracking/_tracking_service/client.py` and remove them from the `Param` and `Metric` entity definitions?

File: mlflow/entities/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/entities/metic_value_conversion_utils.py:39]: @ntakouris I'm still not fully following why we need to special case string conversion for the use of tensors / pyspark mlparams as metric keys (names). I think it should be sufficient to convert all `key` inputs to string using `str()`. Is there a particular use case you have for using a tensor as a metric name?

File: examples/h2o/random_forest.ipynb
Comment (ntakouris) [reply examples/h2o/random_forest.ipynb:94]: Yes, I have not yet pushed all the changes.

File: mlflow/entities/metic_value_conversion_utils.py
Comment (ntakouris) [reply mlflow/entities/metic_value_conversion_utils.py:39]: I've not pushed my local changes.

**Comment by ntakouris** [comment]
Are those the api changes you expect?

File: mlflow/tracking/_tracking_service/client.py
Comment (dbczumar) [inline mlflow/tracking/_tracking_service/client.py:241]: @ntakouris Do we need to make changes here? The string representation of an ndarray is already human-readable:

```
In [4]: k = np.array([1, 2, 3])

In [5]: str(k)
Out[5]: '[1 2 3]'
```

File: mlflow/tracking/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/tracking/metic_value_conversion_utils.py:1]: Nit: Looks like there's a small typo in the naming of this file (missing an `r` in metric)

Can we rename to `metric_value_conversion_utils.py`?

File: mlflow/tracking/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/tracking/metic_value_conversion_utils.py:15]: Can we add the INVALID_PARAMETER_VALUE `error_code` as well? Example: https://github.com/mlflow/mlflow/blob/1539d9e467bcd865c335e800dad597517b0a9bcd/mlflow/pyfunc/__init__.py#L791-L793

File: mlflow/tracking/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/tracking/metic_value_conversion_utils.py:10]: Nit: Can we use single underscore prefixes for these method names?

File: mlflow/tracking/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/tracking/metic_value_conversion_utils.py:18]:
```suggestion
    try:
        return x.item()
    except ValueError:
        raise MlflowException(
            f"Expected metric value to contain a single element, got shape {x.shape} instead"
        )
```

File: mlflow/tracking/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/tracking/metic_value_conversion_utils.py:21]: Very neat decorator! :D

File: mlflow/tracking/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/tracking/metic_value_conversion_utils.py:52]: Following from https://github.com/mlflow/mlflow/pull/5756/files#r866392045, I don't think we strictly need this method at this time. I also see that it's only called for parameter values (not metrics), so, at minimum, we should adjust the naming

File: mlflow/tracking/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/tracking/metic_value_conversion_utils.py:55]: Looks like `convert_int = False` is never specified by any callers. Shall we remove it?

Looks like we always convert to float when logging metrics anyway:

```
import mlflow

from mlflow.tracking.client import MlflowClient

client = MlflowClient()

run = mlflow.start_run()
mlflow.log_metric("M", 1)

run_fetched = client.get_run(run.info.run_id)
print(run_fetched.data.metrics)
```

yields

```
1.0
```

File: mlflow/tracking/metic_value_conversion_utils.py
Comment (dbczumar) [inline mlflow/tracking/metic_value_conversion_utils.py:6]:
```suggestion
def _is_module_imported(module_name: str) -> bool:
```

File: tests/tracking/fluent/test_metric_value_conversion_utils.py
Comment (dbczumar) [inline tests/tracking/fluent/test_metric_value_conversion_utils.py:15]: Nit: Can we remove `unittest.TestCase` in favor of pytest-style test cases? For example, see https://github.com/mlflow/mlflow/blob/master/tests/models/test_model_input_examples.py.

File: tests/tracking/test_tracking.py
Comment (dbczumar) [reply tests/tracking/test_tracking.py:454]: Ideally, it should be up to the backend to determine the validity of a metric value. If our attempted conversion fails, instead of throwing, can we propagate the original metric value as part of the request?

**Review by dbczumar — COMMENTED** [review]
@ntakouris This is looking great! I left a few more comments; we should be close to merge-ready once they're addressed! Thanks a bunch!

File: mlflow/tracking/_tracking_service/client.py
Comment (ntakouris) [reply mlflow/tracking/_tracking_service/client.py:241]: Agreed, will remove.

File: mlflow/tracking/metic_value_conversion_utils.py
Comment (ntakouris) [reply mlflow/tracking/metic_value_conversion_utils.py:10]: Sure thing!

**Comment by ntakouris** [comment]
If latest changes are ok, I can proceed to add a few sanity tests for the high level client logging api.

**Comment by dbczumar** [comment]
> If latest changes are ok, I can proceed to add a few sanity tests for the high level client logging api.

Hi @ntakouris, apologies for the delay! The latest changes look great; some test coverage would be fantastic! Thank you for your contribution!

**Comment by ntakouris** [comment]
@dbczumar @WeichenXu123 Bump, this should be ready to merge

File: tests/tracking/fluent/test_metric_value_conversion_utils.py
Comment (harupy) [inline tests/tracking/fluent/test_metric_value_conversion_utils.py:47]: Can we simplify the code?

```suggestion
    with start_run() as run:
        mlflow.log_metric("name_numpy", ndarray_val)

    finished_run = tracking.MlflowClient().get_run(run.info.run_id)
    assert len(finished_run.data.metrics) == {"name_numpy": ndarray_float_val}
```

File: tests/tracking/test_tracking.py
Comment (harupy) [inline tests/tracking/test_tracking.py:456]: Why do we need this change? `match` should not be removed.

**Comment by ntakouris** [comment]
@harupy Addressed both comments.

Also rebased and force-pushed on master => CI got fixed.

File: tests/pytorch/test_pytorch_metric_value_convertion_utils.py
Comment (harupy) [inline tests/pytorch/test_pytorch_metric_value_convertion_utils.py:8]: Can we avoid using wild car import?

File: tests/pytorch/test_pytorch_metric_value_convertion_utils.py
Comment (harupy) [inline tests/pytorch/test_pytorch_metric_value_convertion_utils.py:2]: Let's remove unused `mock.`

File: tests/pytorch/test_pytorch_metric_value_convertion_utils.py
Comment (harupy) [inline tests/pytorch/test_pytorch_metric_value_convertion_utils.py:16]:
```suggestion
    with pytest.raises(MlflowException, match="<replace me>") as e:
```

Can we add `match`?

File: tests/pytorch/test_pytorch_metric_value_convertion_utils.py
Comment (harupy) [inline tests/pytorch/test_pytorch_metric_value_convertion_utils.py:35]:
```suggestion
    assert finished_run.data.metrics == {"name_torch": torch_tensor_float_val}
```

File: tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py
Comment (harupy) [inline tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py:35]:
```suggestion
    assert finished_run.data.metrics == {"name_tf": tf_tensor_float_val}
```

File: tests/tracking/fluent/test_metric_value_conversion_utils.py
Comment (harupy) [inline tests/tracking/fluent/test_metric_value_conversion_utils.py:18]: same here https://github.com/mlflow/mlflow/pull/5756/files#r889763618

File: tests/tracking/test_tracking.py
Comment (harupy) [inline tests/tracking/test_tracking.py:456]: Let's revert this change.

File: tests/pytorch/test_pytorch_metric_value_convertion_utils.py
Comment (harupy) [inline tests/pytorch/test_pytorch_metric_value_convertion_utils.py:24]:
```suggestion
    assert convert_metric_value_to_float_if_possible(torch_tensor_val) == float(torch_tensor_val[0])
```

Can we move the actual value to the left?

Comment (harupy) [reply tests/pytorch/test_pytorch_metric_value_convertion_utils.py:24]: Can we also apply this change to other assert statements?

**Comment by harupy** [comment]
@ntakouris Left a few more comments. LGTM once they are addressed.

File: tests/pytorch/test_pytorch_metric_value_convertion_utils.py
Comment (ntakouris) [reply tests/pytorch/test_pytorch_metric_value_convertion_utils.py:8]: Of course! will also update other 2 tests that use this.

**Comment by ntakouris** [comment]
Addressed the changes.

File: tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py
Comment (harupy) [inline tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py:24]:
```suggestion
    tf_tensor_val = tf.random.uniform([])  # pytlin: disable=no-value-for-parameter
```

Let's add a `# pytlin: disable=no-value-for-parameter` here to ignore the pylint error below.

File: tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py
Comment (harupy) [inline tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py:24]:
```suggestion
    tf_tensor_val = tf.random.uniform([])  # pylint: disable=no-value-for-parameter
```

Sorry for the typo.

File: tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py
Comment (harupy) [inline tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py:29]:
```suggestion
    tf_tensor_val = tf.random.uniform([])  # pylint: disable=no-value-for-parameter
```

File: tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py
Comment (harupy) [inline tests/tensorflow/test_tensorflow2_metric_value_convertion_utils.py:13]:
```suggestion
    multi_item_tf_tensor = tf.random.uniform([2, 2])  # pylint: disable=no-value-for-parameter
```

**Comment by ntakouris** [comment]
@harupy No need to skip pylint. I just added `dtype=tf.float32`.

**Comment by ntakouris** [comment]
@harupy pylint passes locally. Mind letting this CI go one more time?

File: mlflow/tracking/metric_value_conversion_utils.py
Comment (harupy) [inline mlflow/tracking/metric_value_conversion_utils.py:21]: Let's fix this line.

**Comment by ntakouris** [comment]
@harupy  I reduced the length of that line. It wasn't caught by `black`.

**Comment by ntakouris** [comment]
Looks good now?

File: mlflow/tracking/metric_value_conversion_utils.py
Comment (harupy) [inline mlflow/tracking/metric_value_conversion_utils.py:90]: same here https://github.com/mlflow/mlflow/pull/5756/files#r890268871

**Comment by ntakouris** [comment]
@harupy We now retain the original exception message too.

**Review by harupy — APPROVED** [review]
@ntakouris LGTM, thanks for the contribution!

**Pull request merged**
